---
title: "Lecture 11"
author: "BIOF 339"
date: "November 22, 2017"
output: ioslides_presentation
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      fig.width=4,
                      fig.height=4,
                      comment = '# ',
                      cache=F,
                      warning=F, message=F)
```

# Bioconductor

## Bioconductor

Bioconductor provides tools for the analysis and comprehension of high-throughput genomic data, using R.

+ 1476 packages
+ Covers the bioinformatic pipeline

+ Software
+ Annotation
+ Experiments

# Explore Bioconductor [website](http://www.bioconductor.org)

## Installing Bioconductor packages

This is different from the usual `install.packages`

```{r, eval=F}
source('http://bioconductor.org/biocLite.R')
biocLite(c('Biobase','limma','hgu95av2.db','Biostrings'))
```

## DNA sequences

```{r}
library(Biostrings)
dna <- DNAStringSet(c("AACAT", "GGCGCCT"))
reverseComplement(dna)
```

## DNA sequences

```{r}
library(Biostrings)
data("phiX174Phage")
phiX174Phage
```

## DNA sequences

```{r}
letterFrequency(phiX174Phage, 'GC', as.prob=TRUE)
```


## Data in Bioconductor

The basic structure in a Bioconductor pipeline is the `ExpressionSet`

```{r, eval=T}
library(Biobase)
data("sample.ExpressionSet")
str(sample.ExpressionSet)
```

## Differences with usual R

Instead of storing data in named lists, ExpressionSet objects store
data in **slots**, and we can see what the slots are with `slotNames`:

```{r, eval=T}
slotNames(sample.ExpressionSet)
```

## Differences with usual R

You can access these slots using `@`, instead of the usual `$`:

```{r, eval=T}
sample.ExpressionSet@phenoData
```

## Differences with usual R

However, it's much easier to go with the built-in functions
```{r , eval=T}
pData(sample.ExpressionSet)
```

## Differences with usual R

```{r , eval=T}
head(exprs(sample.ExpressionSet))
```
# Bioconductor ecosystem

## 

```{r, fig.width=9, echo=F}
knitr::include_graphics('lecture12_dat/bioc1.png')
```

<div align = 'right'>
<p>Taken from <a href = "http://bioconductor.org/help/course-materials/2017/OSU/B1_Bioconductor_intro.html", 
target="_blank">Morgan's Bioconductor Tutorial</a></p>
</div>

# A RNA-Seq pipeline ([link](http://bioconductor.org/help/course-materials/2017/OSU/B3_RNASeq_Workflow.html#lab-gene-level-rna-seq-differential-expression))

## Goals

+ Exploratory data analysis
+ Differential expression analysis with [DESeq2](https://bioconductor.org/packages/DESeq2)
+ Visualization

+ We will start after reads have been aligned to a reference genome and reads overlapping known genes have been counted

## The experiment

+ In the experiment, four primary human airway smooth muscle cell lines were treated with 1 micromolar dexamethasone for 18 hours. 
+ For each of the four cell lines, we have a treated and an untreated sample.

## Start with prepared SummarizedExperiement
```{r}
library(airway)
data(airway)
se <- airway
head(assay(se))
```

## Metadata for experiment
```{r}
colData(se)
```

## Genomic ranges over which counting occurred

```{r}
rowRanges(se)
```

## Create a DESeqDataSet

```{r}
library("DESeq2")
dds <- DESeqDataSet(se, design = ~ cell + dex)
dds
```

## Fix factor label

```{r}
dds$dex <- relevel(dds$dex, "untrt")
```

## Run differential expression pipeline

```{r}
dds <- DESeq(dds)
```

## Extracting results

```{r}
(res <- results(dds))
```

## Summarizing results
```{r}
summary(res)
```

## Summarizing results
```{r}
library(tidyverse)
as.data.frame(res) %>% 
  filter(padj < 0.1) %>% 
  arrange(desc(abs(log2FoldChange))) %>% head()
```

## A visualization

```{r, fig.height=3, fig.width=6}
topGene <- rownames(res)[which.min(res$padj)]
data <- plotCounts(dds, gene=topGene, intgroup=c("dex"), returnData=TRUE)
ggplot(data, aes(x = dex, y = count, fill=dex))+
  geom_dotplot(binaxis='y', stackdir='center')+
  scale_y_log10()
```

## Another visualization

```{r, fig.width=7}
plotMA(res, ylim=c(-5,5))
```

# Making a heatmap

## Heatmaps

There are several ways of doing heatmaps in R:

+ [http://sebastianraschka.com/Articles/heatmaps_in_r.html](http://sebastianraschka.com/Articles/heatmaps_in_r.html)
+ [https://plot.ly/r/heatmaps/](https://plot.ly/r/heatmaps/)
+ [http://moderndata.plot.ly/interactive-heat-maps-for-r/](http://moderndata.plot.ly/interactive-heat-maps-for-r/)
+ [http://www.siliconcreek.net/r/simple-heatmap-in-r-with-ggplot2](http://www.siliconcreek.net/r/simple-heatmap-in-r-with-ggplot2)
+ [https://rud.is/b/2016/02/14/making-faceted-heatmaps-with-ggplot2/](https://rud.is/b/2016/02/14/making-faceted-heatmaps-with-ggplot2/)

## Some example data

```{r}
library(Biobase)
data(sample.ExpressionSet)
exdat <-  sample.ExpressionSet
library(limma)
design1 <- model.matrix(~type, data=pData(exdat))
lm1 <- lmFit(exprs(exdat), design1)
lm1 <- eBayes(lm1) # compute linear model for each probeset
geneID <- rownames(topTable(lm1, coef=2, num=100, adjust='none',p.value=0.05))
exdat2 <- exdat[geneID,] # Keep features with p-values < 0.05
exdat2
```

## Heatmaps using `Heatplus`

```{r, eval=F}
source('http://bioconductor.org/biocLite.R')
biocLite('Heatplus')
```

---

```{r, echo=T, fig.width=6, fig.height=6, eval=F}
library(Heatplus)
reg1 <- regHeatmap(exprs(exdat2), legend=2, col=heat.colors, 
                   breaks=-3:3)
plot(reg1)
```

---

```{r, echo=F, fig.width=6, fig.height=6, eval=T}
library(Heatplus)
reg1 <- regHeatmap(exprs(exdat2), legend=2, col=heat.colors, 
                   breaks=-3:3)
plot(reg1)
```


```{r, echo=T, fig.width=6, fig.height=6, eval=F}
library(Heatplus)
reg1 <- regHeatmap(exprs(exdat2), legend=2, col=heat.colors, 
                   breaks=-3:3)
plot(reg1)
```

---

```{r, echo=T, eval=F}
corrdist <- function(x) as.dist(1-cor(t(x)))
hclust.avl <-  function(x) hclust(x, method='average')
reg2 <- regHeatmap(exprs(exdat2), legend=2, col=heat.colors,
                   breaks=-3:3, 
                   dendrogram = list(clustfun=hclust.avl, distfun=corrdist))
plot(reg2)
```

---

```{r, echo=F, eval=T, fig.width=7}
corrdist <- function(x) as.dist(1-cor(t(x)))
hclust.avl <-  function(x) hclust(x, method='average')
reg2 <- regHeatmap(exprs(exdat2), legend=2, col=heat.colors,
                   breaks=-3:3, 
                   dendrogram = list(clustfun=hclust.avl, distfun=corrdist))
plot(reg2)
```

---

```{r, echo=T, eval=F}
ann1 <- annHeatmap(exprs(exdat2), ann=pData(exdat2))
plot(ann1)
```

---

```{r, echo=F, eval=T, fig.width=7, fig.height=6}
ann1 <- annHeatmap(exprs(exdat2), ann=pData(exdat2))
plot(ann1)
```

---

```{r, echo=T, eval=F}
ann1 <- annHeatmap(exprs(exdat2), ann=pData(exdat2))
plot(ann1)
```
```{r, echo=T, eval=F}
ann2 <- annHeatmap(exprs(exdat2), ann=pData(exdat2),
                   cluster = list(cuth=7500,
                                  label=c('Control-like','Case-like')))
plot(ann3)
```

---

```{r, echo=F, eval=T, fig.height=6, fig.width=7}
ann2 <- annHeatmap(exprs(exdat2), ann=pData(exdat2),
                   cluster = list(cuth=7500,
                                  label=c('Control-like','Case-like')))
plot(ann2)
```

---

```{r, fig.height=5, fig.width=6}
library(d3heatmap)
d3heatmap(exprs(exdat2), distfun = corrdist, 
          hclustfun = hclust.avl, scale='row')
```

Put your mouse over each point :)
